conjure-runtime/src/service/gzip.rs

```diff
@@ -0,0 +1,341 @@
+    }
+
+    fn poll_trailers(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<HeaderMap<HeaderValue>>, Self::Error>> {
+        Pin::new(&mut self.body).poll_trailers(cx)
+    }
+
+    fn is_end_stream(&self) -> bool {
+        self.body.is_end_stream()
+    }
+
+    fn size_hint(&self) -> SizeHint {
+        self.body.size_hint()
+    }
+}
+
+impl Stream for DecodedBody {
+    type Item = io::Result<Bytes>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.poll_data(cx)
+    }
+}
+
+#[pin_project]
+struct IdentityBody<T> {
+    #[pin]
+    body: T,
+}
+
+impl<T, E> Body for IdentityBody<T>
+where
+    T: Body<Data = Bytes, Error = E>,
+    E: Into<Box<dyn Error + Sync + Send>>,
+{
+    type Data = Bytes;
+    type Error = io::Error;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        let item = ready!(self.project().body.poll_data(cx));
+        Poll::Ready(item.map(|r| r.map_err(|e| io::Error::new(io::ErrorKind::Other, e))))
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<HeaderMap<HeaderValue>>, Self::Error>> {
+        self.project()
+            .body
+            .poll_trailers(cx)
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
+    }
+
+    fn is_end_stream(&self) -> bool {
+        self.body.is_end_stream()
+    }
+
+    fn size_hint(&self) -> SizeHint {
+        Body::size_hint(&self.body)
+    }
+}
+
+#[pin_project]
+struct ShimStream<T> {
+    #[pin]
+    body: IdentityBody<T>,
+}
+
+impl<T, E> Stream for ShimStream<T>
+where
+    T: Body<Data = Bytes, Error = E>,
+    E: Into<Box<dyn Error + Sync + Send>>,
+{
+    type Item = io::Result<Bytes>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.project().body.poll_data(cx)
+    }
+}
+
+#[pin_project]
+struct GzipBody<T> {
+    #[pin]
+    stream: GzipDecoder<ShimStream<T>>,
+}
+
+impl<T, E> Body for GzipBody<T>
+where
+    T: Body<Data = Bytes, Error = E>,
+    E: Into<Box<dyn Error + Sync + Send>>,
+{
+    type Data = Bytes;
+    type Error = io::Error;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        self.project().stream.poll_next(cx)
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<HeaderMap<HeaderValue>>, Self::Error>> {
+        self.project()
+            .stream
+            .get_pin_mut()
+            .project()
+            .body
+            .poll_trailers(cx)
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
+    }
+
+    // we can't override is_end_stream since we may get an error out of the gzip decoder at eof
```

Comment:
Is this the best place for this comment? How about above [line 216](https://github.com/palantir/conjure-rust-runtime/pull/32/files#diff-17d875a42152ea09fa8c11d259b4dc0eR216)?

Reply:
I stuck it here since that's where the override "would go" but the top of the impl works too.